ttnn/cpp/ttnn/tensor/tensor.cpp

```diff
@@ -450,31 +451,7 @@ bool Tensor::is_allocated() const {
     ZoneScoped;
     auto output = std::visit(
         [](auto&& storage) -> bool {
-            using T = std::decay_t<decltype(storage)>;
-            if constexpr (std::is_same_v<T, OwnedStorage>) {
-                return std::visit([](auto&& buffer) -> bool { return buffer.is_allocated(); }, storage.buffer);
-            } else if constexpr (std::is_same_v<T, DeviceStorage>) {
-                return bool(storage.buffer) and storage.buffer->size() > 0;
-            } else if constexpr (std::is_same_v<T, BorrowedStorage>) {
-                return true;
-            } else if constexpr (std::is_same_v<T, MultiDeviceHostStorage>) {
-                bool is_allocated = true;
-                for (int i = 0; i < storage.num_buffers(); i++) {
-                    is_allocated &=
-                        std::visit([](auto&& buffer) -> bool { return buffer.is_allocated(); }, storage.get_buffer(i));
-                }
-                return is_allocated;
-            } else if constexpr (std::is_same_v<T, MultiDeviceStorage>) {
-                bool is_allocated = true;
-                for (int i = 0; i < storage.ordered_device_ids.size(); ++i) {
-                    auto device_id = storage.ordered_device_ids[i];
-                    const auto& buffer = storage.get_buffer_for_device_id(device_id);
-                    is_allocated &= bool(buffer) and buffer->size() > 0;
-                }
-                return is_allocated;
-            } else {
-                raise_unsupported_storage<T>();
-            }
+            return storage.is_allocated();
         },
```

Comment:
This lambda:
```cpp
[](auto&& storage) -> bool { return storage.is_allocated(); }
```
gets repeated a few times. I think it would be worth defining
```cpp
struct is_allocated_fn {
  constexpr bool operator(const auto &storage) const {
    return storage.is_allocated();
  }
};

inline constexpr is_allocated_fn is_allocated{};
```
and using `is_allocated` in place of these lambdas each time. Would prevent a few unnecessary template instantiations, and consolidate the implementation back into one place in case future refactoring becomes necessary.

Reply:
I don't want to add more complexity to the existing code.
I don't like this lambda but it is just one line of code.

Reply:
Also I'd create just simple function for this case.

Reply:
It's not more complexity, it's a simplification. Rather than having 3 separate lambdas for the same purpose, you have one functor that gets re-used in three places, making it clear with an identifier that their implementation is purposefully. But just to convince you more, you could even write this:
```cpp
struct is_allocated_fn {
  constexpr bool operator(const auto &storage_alternatives) const {
    return std::visit([](const auto &storage) { return storage.is_allocated(); },
                      storage_alternatives);
  }
};

inline constexpr is_allocated_fn is_allocated{};
```
and refactor the 3 occurrences of the `std::visit` call passing the above lambda with simply:
```cpp
auto output = is_allocated(this->get_storage());
```